File: golang_archive/pkg/ikuai_api4/custom_isp.go
package ikuai_api4

import (
	"errors"
	"ikuai-bypass/pkg/ikuai_common"
	"strconv"
	"strings"
)

const FUNC_NAME_CUSTOM_ISP = "custom_isp"

// ShowCustomIspByTagName 根据标签名称查询自定义运营商规则
// ShowCustomIspByTagName queries custom ISP rules by tag name
func (i *IKuai) ShowCustomIspByTagName(tagName string) (result []ikuai_common.CustomIspData, err error) {
	param := struct {
		Type    string `json:"TYPE"`
		Limit   string `json:"limit"`
		OrderBy string `json:"ORDER_BY"`
		Order   string `json:"ORDER"`
	}{
		Type:  "total,data",
		Limit: "0,1000",
	}
	req := CallReq{
		FuncName: FUNC_NAME_CUSTOM_ISP,
		Action:   "show",
		Param:    &param,
	}
	var all []ikuai_common.CustomIspData
	resp := CallResp{Results: &CallRespData{Data: &all}}
	err = postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return
	}
	if resp.Code != 0 {
		err = errors.New(resp.Message)
		return
	}
	for _, d := range all {
		if matchTagNameFilter(tagName, d.Name, d.Comment) {
			result = append(result, d)
		}
	}
	return
}

// AddCustomIsp 添加自定义运营商规则
// AddCustomIsp adds a custom ISP rule
func (i *IKuai) AddCustomIsp(tag, ipgroup string, index int) error {
	ipgroup = strings.TrimSpace(ipgroup)

	param := struct {
		Name    string `json:"name"`
		Ipgroup string `json:"ipgroup"`
		Comment string `json:"comment"`
	}{
		Name:    buildTagName(tag),
		Ipgroup: ipgroup,
		Comment: buildCustomIspChunkComment(index),
	}
	req := CallReq{
		FuncName: FUNC_NAME_CUSTOM_ISP,
		Action:   "add",
		Param:    &param,
	}
	resp := CallResp{}
	err := postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return err
	}
	if resp.Code != 0 {
		return errors.New(resp.Message)
	}
	return nil
}

func (i *IKuai) EditCustomIsp(tag, ipgroup string, index int, id int) error {
	ipgroup = strings.TrimSpace(ipgroup)

	param := struct {
		Name    string `json:"name"`
		Ipgroup string `json:"ipgroup"`
		Comment string `json:"comment"`
		ID      int    `json:"id"`
	}{
		Name:    buildTagName(tag),
		Ipgroup: ipgroup,
		Comment: buildCustomIspChunkComment(index),
		ID:      id,
	}
	req := CallReq{
		FuncName: FUNC_NAME_CUSTOM_ISP,
		Action:   "edit",
		Param:    &param,
	}
	resp := CallResp{}
	err := postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return err
	}
	if resp.Code != 0 {
		return errors.New(resp.Message)
	}
	return nil
}

func (i *IKuai) GetCustomIspMap(tag string) (result map[int]int, err error) {
	result = make(map[int]int)
	var data []ikuai_common.CustomIspData
	data, err = i.ShowCustomIspByTagName("")
	if err != nil {
		return nil, err
	}

	for _, d := range data {
		if matchTagNameFilter(tag, d.Name, d.Comment) {
			idx, ok := parseCustomIspChunkIndexFromComment(d.Comment)
			if !ok {
				idx, ok = parseCustomIspChunkIndexFromName(d.Name, tag)
			}
			if !ok {
				continue
			}
			if existedID, existed := result[idx]; existed {
				i.L.Warn("QUERY:序号冲突", "Duplicate custom ISP chunk %d for tag %s (keep ID: %d, skip ID: %d)", idx, tag, existedID, d.ID)
				continue
			}
			result[idx] = d.ID
		}
	}
	return result, nil
}

func buildCustomIspChunkComment(index int) string {
	chunk := index + 1
	if chunk <= 1 {
		return ikuai_common.NEW_COMMENT
	}
	return ikuai_common.NEW_COMMENT + "-" + strconv.Itoa(chunk)
}

func parseCustomIspChunkIndexFromComment(comment string) (int, bool) {
	comment = strings.TrimSpace(comment)
	if comment == "" {
		return 0, false
	}

	prefixes := []string{ikuai_common.NEW_COMMENT, ikuai_common.COMMENT_IKUAI_BYPASS}
	for _, prefix := range prefixes {
		if comment == prefix {
			return 1, true
		}
		if !strings.HasPrefix(comment, prefix) {
			continue
		}

		suffix := strings.TrimPrefix(comment, prefix)
		suffix = strings.TrimPrefix(suffix, "-")
		suffix = strings.TrimPrefix(suffix, "_")
		suffix = strings.TrimSpace(suffix)
		if suffix == "" {
			return 1, true
		}

		idx, err := strconv.Atoi(suffix)
		if err == nil && idx > 0 {
			return idx, true
		}
	}

	return 0, false
}

func parseCustomIspChunkIndexFromName(name, tag string) (int, bool) {
	baseName := buildTagName(tag)
	suffix := strings.TrimPrefix(strings.TrimSpace(name), baseName)
	if suffix == "" {
		return 0, false
	}
	idx, err := strconv.Atoi(suffix)
	if err != nil || idx <= 0 {
		return 0, false
	}
	return idx, true
}

// DelCustomIsp 删除自定义运营商规则
// DelCustomIsp deletes a custom ISP rule
func (i *IKuai) DelCustomIsp(id string) error {
	param := struct {
		Id string `json:"id"`
	}{
		Id: id,
	}
	req := CallReq{
		FuncName: FUNC_NAME_CUSTOM_ISP,
		Action:   "del",
		Param:    &param,
	}
	resp := CallResp{}
	err := postJson(i.client, i.baseurl+"/Action/call", &req, &resp)
	if err != nil {
		return err
	}
	if resp.Code != 0 {
		return errors.New(resp.Message)
	}
	return nil
}

// DelCustomIspAll 清理所有符合条件的自定义运营商规则
// DelCustomIspAll cleans up all custom ISP rules that match the condition
func (i *IKuai) DelCustomIspAll(cleanTag string) (err error) {
	for {
		var data []ikuai_common.CustomIspData
		data, err = i.ShowCustomIspByTagName("")
		if err != nil {
			return
		}
		var ids []string
		for _, d := range data {
			if matchCleanTag(cleanTag, d.Comment, d.Name) {
				ids = append(ids, strconv.Itoa(d.ID))
			}
		}
		if len(ids) <= 0 {
			return
		}
		id := strings.Join(ids, ",")
		err = i.DelCustomIsp(id)
		if err != nil {
			return
		}
	}
}

File: core/src/ikuai/custom_isp.rs
use serde::Serialize;

use super::clean::match_clean_tag;
use super::tag_name::{build_tag_name, match_tag_name_filter};
use super::types::{CustomIspData, IKuaiClient, IKuaiError, FUNC_NAME_CUSTOM_ISP, NEW_COMMENT};

#[derive(Debug, Serialize)]
struct ShowParam {
    #[serde(rename = "TYPE")]
    r#type: String,
    limit: String,
}

#[derive(Debug, Serialize)]
struct DelParam {
    id: String,
}

pub async fn show_custom_isp_by_tag_name(
    api: &IKuaiClient,
    tag_name: &str,
) -> Result<Vec<CustomIspData>, IKuaiError> {
    let param = ShowParam {
        r#type: "total,data".to_string(),
        limit: "0,1000".to_string(),
    };
    let resp = api
        .call::<_, Vec<CustomIspData>>(FUNC_NAME_CUSTOM_ISP, "show", &param)
        .await?;
    let data = resp
        .results
        .ok_or_else(|| IKuaiError::InvalidResponse("missing results".to_string()))?
        .data;
    Ok(data
        .into_iter()
        .filter(|d| match_tag_name_filter(tag_name, &d.name, &d.comment))
        .collect())
}

pub async fn add_custom_isp(api: &IKuaiClient, tag: &str, ipgroup: &str, index: i64) -> Result<(), IKuaiError> {
    let param = serde_json::json!({
        "name": build_tag_name(tag),
        "ipgroup": ipgroup.trim(),
        "comment": build_custom_isp_chunk_comment(index),
    });
    let _ = api
        .call::<_, serde_json::Value>(FUNC_NAME_CUSTOM_ISP, "add", &param)
        .await?;
    Ok(())
}

pub async fn edit_custom_isp(
    api: &IKuaiClient,
    tag: &str,
    ipgroup: &str,
    index: i64,
    id: i64,
) -> Result<(), IKuaiError> {
    let param = serde_json::json!({
        "name": build_tag_name(tag),
        "ipgroup": ipgroup.trim(),
        "comment": build_custom_isp_chunk_comment(index),
        "id": id,
    });
    let _ = api
        .call::<_, serde_json::Value>(FUNC_NAME_CUSTOM_ISP, "edit", &param)
        .await?;
    Ok(())
}

pub async fn del_custom_isp(api: &IKuaiClient, id_csv: &str) -> Result<(), IKuaiError> {
    let param = DelParam {
        id: id_csv.to_string(),
    };
    let _ = api
        .call::<_, serde_json::Value>(FUNC_NAME_CUSTOM_ISP, "del", &param)
        .await?;
    Ok(())
}

pub async fn get_custom_isp_map(api: &IKuaiClient, tag: &str) -> Result<std::collections::HashMap<i64, i64>, IKuaiError> {
    let data = show_custom_isp_by_tag_name(api, "").await?;
    let mut out = std::collections::HashMap::new();
    for d in data {
        if !match_tag_name_filter(tag, &d.name, &d.comment) {
            continue;
        }
        let idx = parse_custom_isp_chunk_index_from_comment(&d.comment)
            .or_else(|| parse_custom_isp_chunk_index_from_name(&d.name, tag));
        let Some(idx) = idx else { continue };
        out.entry(idx).or_insert(d.id);
    }
    Ok(out)
}

pub async fn del_custom_isp_all(api: &IKuaiClient, clean_tag: &str) -> Result<(), IKuaiError> {
    loop {
        let data = show_custom_isp_by_tag_name(api, "").await?;
        let ids: Vec<String> = data
            .into_iter()
            .filter(|d| match_clean_tag(clean_tag, &d.comment, &d.name))
            .map(|d| d.id.to_string())
            .collect();
        if ids.is_empty() {
            return Ok(());
        }
        del_custom_isp(api, &ids.join(",")).await?;
    }
}

fn build_custom_isp_chunk_comment(index: i64) -> String {
    let chunk = index + 1;
    if chunk <= 1 {
        return NEW_COMMENT.to_string();
    }
    format!("{}-{}", NEW_COMMENT, chunk)
}

fn parse_custom_isp_chunk_index_from_comment(comment: &str) -> Option<i64> {
    let c = comment.trim();
    if c.is_empty() {
        return None;
    }
    for prefix in [NEW_COMMENT, super::types::COMMENT_IKUAI_BYPASS] {
        if c == prefix {
            return Some(1);
        }
        if !c.starts_with(prefix) {
            continue;
        }
        let mut suffix = c.trim_start_matches(prefix);
        suffix = suffix.trim_start_matches('-');
        suffix = suffix.trim_start_matches('_');
        let suffix = suffix.trim();
        if suffix.is_empty() {
            return Some(1);
        }
        if let Ok(v) = suffix.parse::<i64>() && v > 0 {
            return Some(v);
        }
    }
    None
}

fn parse_custom_isp_chunk_index_from_name(name: &str, tag: &str) -> Option<i64> {
    let base = build_tag_name(tag);
    let suffix = name.trim().trim_start_matches(&base);
    if suffix.is_empty() {
        return None;
    }
    let v = suffix.parse::<i64>().ok()?;
    if v <= 0 {
        return None;
    }
    Some(v)
}
